### ISSUE TITLE
 When using yolo export to convert a model to ONNX format, a decrease in accuracy is observed

### ISSUE THREAD

**Issue by xla145** [ultralytics/ultralytics#10289]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

After training a custom dataset using YOLOv8 and outputting a .pt model, I exported it to ONNX using yolo export, but found that the accuracy was significantly reduced compared to the original .pt model. Figure 1 shows the results using yolo predict, and Figure 2 shows the results using the official Python code for ONNX inference. The recognition performance in Figure 2 is noticeably worse.
![AX8428119_1_1_185680_20240423100431](https://github.com/ultralytics/ultralytics/assets/18521352/2999da30-468b-4c38-bba3-6fd9c3e6673c)
![Output](https://github.com/ultralytics/ultralytics/assets/18521352/bad02ecf-56e8-4eb9-b543-6e642f30d672)

**Comment by glenn-jocher** [comment]
@xla145 hello there! 👋

Thank you for bringing this issue to our attention. It's not uncommon to experience discrepancies in model performance when transitioning between formats, such as from .pt to ONNX. This can happen due to various factors including differences in the handling of operations between the frameworks, which might lead to slight divergences in inference results.

A few steps you might consider to mitigate this issue include:

- **Ensuring Consistency:** Double-check that the preprocessing and postprocessing steps are consistent between both your .pt and ONNX model inferences. Inconsistencies here are often culprits for performance differences.
  
- **Model Optimizations:** When exporting models to ONNX, some optimization choices might affect the model's fidelity. If you're using any optimization flags during export, you might want to try without them and compare the results.

- **Version Compatibility:** Ensure that the ONNX model is being executed with a compatible runtime version. Mismatches can sometimes introduce unexpected behaviors.

- **Dynamic vs. Static Axes:** If your model uses dynamic axes during export (`dynamic=True`), verify that these settings align with your actual inference scenarios. Inconsistencies here might affect performance.

Here's a simple example of a consistent preprocessing step to ensure accuracy during inference:
```python
import torch
from PIL import Image
from torchvision.transforms import functional as F

# Load an image
image = Image.open('path_to_your_image.jpg').convert('RGB')

# Preprocessing steps
image = F.to_tensor(image)
image = F.resize(image, [640, 640])
image = torch.unsqueeze(image, 0)  # Add batch dimension

# Proceed with ONNX model inference here
```

Should the issue persist, consider sharing your specific export command and any additional code snippets used for performing inference in both cases. This could help further diagnose the problem. 

We appreciate your contribution to making YOLOv8 better, and look forward to resolving this together!

**Comment by Dapid** [comment]
You should expect a small degradation in performance because of some quantisation. But your scores are over 0.8, so it should be fine.

Just to make sure, can you run it again with a lower confidence threshold? Do you see hats being detected in the same spots, even if they have very low scores?

Also, let me guess, checking the workers wear hard hats? ;-)

**Comment by glenn-jocher** [comment]
@Dapid Hi there! 😊

Absolutely, a slight performance dip due to quantization is indeed something we occasionally encounter. The scores you're getting do sound promising though!

Trying out a lower confidence threshold is a great idea! It'll help us see if the model still recognizes those hard hats, just with reduced confidence. Here's a quick code snippet on how to adjust that for ONNX inference:

```python
results = model(image, conf_thres=0.25)  # Lowering the confidence threshold to 0.25
```

And yes, you guessed it right! It's indeed about ensuring safety through hard hat detection. Love your keen observation! 🏗️ Let me know how it goes with the adjusted threshold, looking forward to your findings!

**Comment by xla145** [comment]
First of all, thank you for helping me clarify my doubts. I have attempted to make some modifications and used the official demo to test this issue. Below are the steps I took for testing:

1. The command to export a model in ONNX format using the official method is:

!yolo export model=../models/best_hat_20240424_14000.pt format=onnx nms=true device=0

2. The code for processing the .pt model for inference is as follows

import os
from ultralytics import YOLO
if __name__ == '__main__':
    model_path = "xxx"
    model = YOLO(os.path.join(model_path, "best_hat_20240424_14000.pt"))
    results = model.predict(source="AX8428119_1_1_185680_20240423100431.png", save=True, conf=0.5, iou=0.5)

3. code for processing the ONNX model using the official Python demo

```
# Ultralytics YOLO 🚀, AGPL-3.0 license
import os

import cv2
import numpy as np
import onnxruntime as ort

class YOLOv8:
    """YOLOv8 object detection model class for handling inference and visualization."""

    def __init__(self, onnx_model, confidence_thres, iou_thres):
        """
        Initializes an instance of the YOLOv8 class.

        Args:
            onnx_model: Path to the ONNX model.
            input_image: Path to the input image.
            confidence_thres: Confidence threshold for filtering detections.
            iou_thres: IoU (Intersection over Union) threshold for non-maximum suppression.
        """
        self.onnx_model = onnx_model
        self.confidence_thres = confidence_thres
        self.iou_thres = iou_thres

        # Load the class names from the COCO dataset
        self.classes = ['no_hat', 'not_life_jacket', 'not_gloves']

        # Generate a color palette for the classes
        self.color_palette = np.random.uniform(0, 255, size=(len(self.classes), 3))

    def draw_detections(self, img, box, score, class_id):
        """
        Draws bounding boxes and labels on the input image based on the detected objects.

        Args:
            img: The input image to draw detections on.
            box: Detected bounding box.
            score: Corresponding detection score.
            class_id: Class ID for the detected object.

        Returns:
            None
        """

        # Extract the coordinates of the bounding box
        x1, y1, w, h = box

        # Retrieve the color for the class ID
        color = self.color_palette[class_id]

        # Draw the bounding box on the image
        cv2.rectangle(img, (int(x1), int(y1)), (int(x1 + w), int(y1 + h)), color, 2)

        # Create the label text with class name and score
        label = f"{self.classes[class_id]}: {score:.2f}"

        # Calculate the dimensions of the label text
        (label_width, label_height), _ = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)

        # Calculate the position of the label text
        label_x = x1
        label_y = y1 - 10 if y1 - 10 > label_height else y1 + 10

        # Draw a filled rectangle as the background for the label text
        cv2.rectangle(
            img, (label_x, label_y - label_height), (label_x + label_width, label_y + label_height), color, cv2.FILLED
        )

        # Draw the label text on the image
        cv2.putText(img, label, (label_x, label_y), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 0), 1, cv2.LINE_AA)

    def preprocess(self, input_image):
        """
        Preprocesses the input image before performing inference.

        Returns:
            image_data: Preprocessed image data ready for inference.
        """
        # Read the input image using OpenCV
        # self.img = cv2.imread(input_image)
        self.img = input_image

        # Get the height and width of the input image
        self.img_height, self.img_width = self.img.shape[:2]

        # Convert the image color space from BGR to RGB
        img = cv2.cvtColor(self.img, cv2.COLOR_BGR2RGB)

        # Resize the image to match the input shape
        img = cv2.resize(img, (self.input_width, self.input_height))

        # Normalize the image data by dividing it by 255.0
        image_data = np.array(img) / 255.0

        # Transpose the image to have the channel dimension as the first dimension
        image_data = np.transpose(image_data, (2, 0, 1))  # Channel first

        # Expand the dimensions of the image data to match the expected input shape
        image_data = np.expand_dims(image_data, axis=0).astype(np.float32)

        # Return the preprocessed image data
        return image_data

    def postprocess(self, input_image, output, is_draw=False):
        """
        Performs post-processing on the model's output to extract bounding boxes, scores, and class IDs.

        Args:
            input_image (numpy.ndarray): The input image.
            output (numpy.ndarray): The output of the model.

        Returns:
            numpy.ndarray: The input image with detections drawn on it.
        """

        # Transpose and squeeze the output to match the expected shape
        outputs = np.transpose(np.squeeze(output[0]))

        # Get the number of rows in the outputs array
        rows = outputs.shape[0]

        # Lists to store the bounding boxes, scores, and class IDs of the detections
        boxes = []
        scores = []
        class_ids = []

        # Calculate the scaling factors for the bounding box coordinates
        x_factor = self.img_width / self.input_width
        y_factor = self.img_height / self.input_height

        # Iterate over each row in the outputs array
        for i in range(rows):
            # Extract the class scores from the current row
            classes_scores = outputs[i][4:]

            # Find the maximum score among the class scores
            max_score = np.amax(classes_scores)

            # If the maximum score is above the confidence threshold
            if max_score >= self.confidence_thres:
                # Get the class ID with the highest score
                class_id = np.argmax(classes_scores)

                # Extract the bounding box coordinates from the current row
                x, y, w, h = outputs[i][0], outputs[i][1], outputs[i][2], outputs[i][3]

                # Calculate the scaled coordinates of the bounding box
                left = int((x - w / 2) * x_factor)
                top = int((y - h / 2) * y_factor)
                width = int(w * x_factor)
                height = int(h * y_factor)

                # Add the class ID, score, and box coordinates to the respective lists
                class_ids.append(class_id)
                scores.append(max_score)
                boxes.append([left, top, width, height])

        # Apply non-maximum suppression to filter out overlapping bounding boxes
        indices = cv2.dnn.NMSBoxes(boxes, scores, self.confidence_thres, self.iou_thres)

        # Iterate over the selected indices after non-maximum suppression
        _results = []
        for i in indices:
            # Get the box, score, and class ID corresponding to the index
            box = boxes[i]
            score = scores[i]
            class_id = class_ids[i]
            if is_draw:
                # Draw the detection on the input image
                self.draw_detections(input_image, box, score, class_id)
            else:
                _results.append((box, score, class_id))
        # Return the modified input image
        if is_draw:
            return input_image
        else:
            return _results

    def predict(self, input_image, is_draw=False):
        """
        Performs inference using an ONNX model and returns the output image with drawn detections.

        Returns:
            output_img: The output image with drawn detections.
        """
        # Create an inference session using the ONNX model and specify execution providers
        session = ort.InferenceSession(self.onnx_model, providers=["CUDAExecutionProvider", "CPUExecutionProvider"])
        # Get the model inputs
        model_inputs = session.get_inputs()
        # Store the shape of the input for later use
        input_shape = model_inputs[0].shape
        self.input_width = input_shape[2]
        self.input_height = input_shape[3]
        # Preprocess the image data
        img_data = self.preprocess(input_image)
        # Run inference using the preprocessed image data
        outputs = session.run(None, {model_inputs[0].name: img_data})
        # Perform post-processing on the outputs to obtain output image.
        return self.postprocess(self.img, outputs, is_draw)  # output image


if __name__ == "__main__":
    model_path = "xxxx"
    detection = YOLOv8(os.path.join(model_path, "best_hat_20240424_14000.onnx"), 0.5, 0.5)
    img = "AX8428119_1_1_185680_20240423100431.png"
    is_draw = True
    img = cv2.imread(img)
    # Perform object detection and obtain the output image
    _results = detection.predict(img, is_draw=is_draw)
    if is_draw:
        # Display the output image in a window
        cv2.namedWindow("Output", cv2.WINDOW_NORMAL)
        cv2.imshow("Output", _results)
        cv2.imwrite("Output.jpg", _results)
        # Wait for a key press to exit
        cv2.waitKey(0)
```

conf, iou, both before and after images are parameters consistent, but the inference results are inconsistent, with the ONNX model clearly exhibiting lower accuracy，

**Comment by glenn-jocher** [comment]
@xla145 Thank you for the detailed follow-up and for sharing the steps you've taken to test the model exporting and inference process. The thorough comparison and test setup you've provided help us understand the situation better.

Your observation about the ONNX model showing lower accuracy compared to the original PyTorch model is indeed intriguing. This can result from several factors but primarily could be due to differences in processing or operations handling between PyTorch and ONNX runtime environments. 

Given the consistency in both `conf` and `iou` thresholds across your tests, we should consider a few steps to narrow down the potential causes:

1. **Model Conversion Verification:** Double-check the conversion process and ensure that all model layers and operations are supported and correctly mapped by the ONNX format. Sometimes, specific custom operations in PyTorch may not have a direct equivalent in ONNX, which can lead to discrepancies.
2. **Preprocessing and Postprocessing:** Ensure that the input preprocessing (normalization, resizing) and the handling of output tensors (decoding of bounding boxes, applying NMS) are exactly the same for both PyTorch and ONNX inferences.
3. **Runtime Environment:** If possible, test the ONNX model across different runtime environments (e.g., ONNX Runtime, OpenVINO) to check if the issue persists. This could help identify if the problem lies within a specific runtime's handling of the ONNX model.

For further analysis, it might be helpful to inspect the ONNX model's intermediate outputs to ensure they align with what's expected based on the PyTorch model's corresponding outputs. Tools like Netron can visualize ONNX models, and writing custom scripts to compare specific layers' outputs between PyTorch and ONNX could uncover discrepancies.

Lastly, consider reaching out on ONNX community forums or raising an issue in the ONNX GitHub repo if you suspect the conversion issue may be due to unsupported operations or bugs in the conversion process.

Keep experimenting, and let's hope for a resolution soon! If you have more details or updates, feel free to share. 😊

**Comment by Dapid** [comment]
Ultralytics can read onnx directly (pass the path to YOLO as you'd do with the .pt), it is worth a try to make sure you are doing the exact same pre and post processing.

**Comment by glenn-jocher** [comment]
@Dapid Hello! 😊 Yes, you're absolutely right. You can indeed load ONNX models directly with Ultralytics YOLO, just like you would with a .pt model. This ensures consistency in preprocessing and postprocessing steps. Here’s how you can do it:

```python
from ultralytics import YOLO

# Load your ONNX model directly
model = YOLO('path/to/your_model.onnx')

# Now you can use it for inference similarly to a .pt model
results = model.predict(source='path/to/image_or_video.jpg')
```

This approach can help ensure that any differences observed are not due to processing steps. Hope this helps! If you have any further questions or observations, feel free to share. Happy to help!

**Comment by YoulQey** [comment]
I also encountered the same problem. After using YOLOv8 to convert to ONNX format, I predicted everything to be normal with YOLO, but when using CV2 to process the ONNX model, I could hardly detect anything;
After searching for a large amount of information later, it was found that the reason was that the YOLO8 model defaults to RGB space, while the CV2 model defaults to BGR. I hope I can help you

**Comment by glenn-jocher** [comment]
Hello @YoulQey! 😊

Thank you for sharing your experience and insights. It's great to see the community coming together to troubleshoot and resolve issues.

Indeed, the color space difference between RGB (used by YOLO) and BGR (used by OpenCV) can often lead to such discrepancies. Ensuring that the image is correctly converted to the expected color space before inference is crucial.

For anyone encountering similar issues, here's a quick reminder and a code snippet to help ensure the color space is correctly handled:

### Converting BGR to RGB in OpenCV

When using OpenCV to read images for ONNX model inference, make sure to convert the image from BGR to RGB:

```python
import cv2
import numpy as np
import onnxruntime as ort

# Load the image using OpenCV
image_path = 'path/to/your/image.jpg'
image = cv2.imread(image_path)

# Convert the image from BGR to RGB
image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

# Preprocess the image as required by your model
# (e.g., resizing, normalization, etc.)
# ...

# Create an ONNX inference session
session = ort.InferenceSession('path/to/your_model.onnx')

# Run inference
outputs = session.run(None, {session.get_inputs()[0].name: image_rgb})

# Post-process the outputs as needed
# ...
```

### Ensuring Consistency in Preprocessing

To ensure that the preprocessing steps are consistent between the .pt and ONNX models, you can also use Ultralytics to load and run the ONNX model directly, as mentioned earlier:

```python
from ultralytics import YOLO

# Load your ONNX model directly with Ultralytics
model = YOLO('path/to/your_model.onnx')

# Run inference
results = model.predict(source='path/to/image_or_video.jpg')

# Display or save the results
results.show()
results.save('output.jpg')
```

This approach helps maintain consistency in preprocessing and postprocessing, reducing the chances of discrepancies.

If you continue to experience issues, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, providing a minimum reproducible example would be immensely helpful for us to investigate further. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Thank you again for your valuable input, and happy coding! 🚀
